File: plugins/harness/skills/harness-install/assets/uv/runtime/rules/require_single_top_level_kotlin_declaration.py
#!/usr/bin/env -S uv run
# -*- coding: utf-8 -*-
"""Require single top level kotlin declaration rule."""

from collections.abc import Iterable

import re
from pathlib import Path

from harness_check_rule import Finding, HarnessCheckRule

from ._utils import is_safe_directory, read_text, relative, safe_walk, severity_for

class RequireSingleTopLevelKotlinDeclarationRule(HarnessCheckRule):
    """Validate requireSingleTopLevelKotlinDeclaration check."""

    category = "requireSingleTopLevelKotlinDeclaration"

    def applies(self, manifest: dict) -> bool:
        """Check if this rule applies to the manifest."""
        section = manifest.get(self.category)
        if not isinstance(section, dict):
            return False
        enabled = section.get("enabled", True)
        return enabled is not False

    def validate(self, root: Path, manifest: dict) -> Iterable[Finding]:
        """Validate requireSingleTopLevelKotlinDeclaration check."""
        section = manifest.get(self.category, {})
        if not isinstance(section, dict):
            return []
        params = section.get("parameters", {})
        if not isinstance(params, dict):
            return []
        directories = params.get("directories", [])
        if not isinstance(directories, list):
            return []
        messages = section.get("messages", {})
        if not isinstance(messages, dict):
            return []
        template = messages.get(
            "default",
            "kotlin file must have exactly one top-level declaration: {file}",
        )
        top_level_pattern = re.compile(
            r"^(class|interface|enum class|data class|sealed class|object|abstract class|val|var|fun|typealias)\b",
            re.MULTILINE,
        )
        result = []
        for directory in directories:
            if not isinstance(directory, str):
                continue
            dir_path = root / directory
            if not is_safe_directory(dir_path):
                continue
            for path in safe_walk(dir_path):
                if path.suffix != ".kt":
                    continue
                text = read_text(path)
                declaration_count = len(top_level_pattern.findall(text))
                if declaration_count != 1:
                    result.append(Finding(
                        severity_for(manifest, self.category),
                        self.category,
                        template.format(file=relative(path)),
                    ))
        return result


RULE: HarnessCheckRule = RequireSingleTopLevelKotlinDeclarationRule()

File: plugins/harness/skills/harness-install/assets/bun/runtime/rules/require-single-top-level-kotlin-declaration.ts
#!/usr/bin/env bun
import type { Finding, HarnessCheckRule, HarnessManifest, RuleContext } from "../harness-check-rule";

/**
 * Require Kotlin files to have exactly one top-level declaration.
 */
export const requireSingleTopLevelKotlinDeclarationRule = (ctx: RuleContext): HarnessCheckRule => ({
  applies(manifest: HarnessManifest): boolean {
    const section = manifest.requireSingleTopLevelKotlinDeclaration;
    if (typeof section !== "object" || section === null) {
      return false;
    }
    const enabled = (section as { enabled?: unknown }).enabled;
    if (enabled === false) {
      return false;
    }
    const parameters = ctx.readJsonObject((section as Record<string, unknown>).parameters);
    return typeof parameters.directories === "object" && parameters.directories !== null;
  }

  validate(_root: string, manifest: HarnessManifest): readonly Finding[] {
    const parameters = ctx.readJsonObject(ctx.readJsonObject(manifest.requireSingleTopLevelKotlinDeclaration).parameters);
    const directories = Array.isArray(parameters.directories) ? parameters.directories : [];
    const directoryStrs = directories.filter((item): item is string => typeof item === "string");
    return directoryStrs.flatMap((directory) => {
      const [files, warnings] = ctx.walkDirectory(directory);
      return warnings.concat(
        files.flatMap((file) => {
          if (!file.endsWith(".kt")) {
            return [];
          }
          const text = ctx.read(file);
          const declRegex = /^(class|interface|enum class|object|data class|sealed class|val|var|fun|typealias)\s/gm;
          let match: RegExpExecArray | null;
          let count = 0;
          while ((match = declRegex.exec(text)) !== null) {
            count++;
          }
          return count !== 1
            ? [
                {
                  severity: ctx.severityOf(manifest, "requireSingleTopLevelKotlinDeclaration"),
                  category: "requireSingleTopLevelKotlinDeclaration",
                  message: `Kotlin file must have exactly 1 top-level declaration: ${file} (found ${count})`,
                },
              ]
            : [];
        })
      );
    });
  }

});
